.github/workflows/publish.yml

```diff
@@ -0,0 +1,47 @@
+name: Publish to npm
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - 'package.json'
+jobs:
+  test:
+    uses: ./.github/workflows/test.yml
+  get-version:
+    runs-on: ubuntu-latest
+    outputs:
+      current_version: ${{ steps.current_version.outputs.version }}
+      last_version: ${{ steps.last_version.outputs.version }}
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 50
+      # Check if the package.json version field has changed since the last push
+      - name: Get current version from package.json
+        id: current_version
+        run: |
+          echo "version=$(jq -r .version package.json)" >> $GITHUB_OUTPUT
+      - name: Get the version from the last push
+        id: last_version
+        run: |
+          git checkout ${{ github.event.before }}
+          echo "version=$(jq -r .version package.json)" >> $GITHUB_OUTPUT
+  npm-publish:
+    needs:
+      - test
+      - get-version
+    runs-on: ubuntu-latest
+    # We only want to publish if the package.json version field has changed
+    if: needs.get-version.outputs.current_version != needs.get-version.outputs.last_version
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 19
+          cache: 'npm'
```

Comment:
```suggestion
          cache: 'npm'
          always-auth: true
```

The CI fails unless adding this line: https://github.com/Dirain1700/workflows-test/actions/runs/4520306972



Have you tested these CIs? Follow my example CI and change it

[Actions link](https://github.com/Dirain1700/workflows-test/actions/runs/4520384963)

[Auto-commit link](https://github.com/Dirain1700/workflows-test/commit/73b7a77a6fea7727fda1f7c675cf5ff183fd1733)

Reply:
I think that's failing due to the NODE_AUTH_TOKEN not being set though. Do you have a NPM_TOKEN set in that repository you're testing it in? Otherwise publish will never work
https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages

You are correct though that that always-auth is necessary, but as per the github actions docs, adding a `registry-url` should be all that's necessary

This workflow is pretty much identical to the docs there

Reply:
Of course, I had a secret named NPM_TOKEN on that repo (if what you said is true, then my CI must have failed)

anyaways, there is nothing wrong with having authority, i will recommend this.

Reply:
Could you clarify what you mean by having authority? Do you mean the argument to setup node?

Reply:
In fact, your prediction is off. You should specify explicitly
![2D46E0A3-1BFB-48AC-A9F5-DCE7815726F5](https://user-images.githubusercontent.com/86363603/227758201-4616c0eb-da46-4ce5-8d27-6ffd360892a4.jpeg)

Reply:
Oh yeah, you're 100% correct. If ommited it does indeed set that flag to false
![image](https://user-images.githubusercontent.com/6253896/227775951-c9eb8c13-4704-40f7-9d44-a179e1df6296.png)

I just want to check though, are you sure that tag is necessary? Every single online sample GA npm publish workflow I've found does not state that option.

Why is it required? Do we know _why_ it's failing without it?

Reply:
that image is your script and failing, i specified that on my script

Reply:
Aye, however I've since added the `registry-url` which was missing from mine, could you edit yours to include that option? From what I've been searching, the always-auth option shouldn't be necessary, just the `registry-url` which I forgot to add

Reply:
Yeah, `always-auth` is not needed, thats my bad.

Reply:
No problem! Yeah I forgot the registry in the first place so that probably would have fixed it instead